### PR TITLE
cart-code endpoint

### PR DESCRIPTION
**Pull request by khushal-rathore-metacube**
## PR Description

CC-16781 : cart-code endpoint

### CODE REVIEW

File: tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot [hunk 1]
```diff
@@ -0,0 +1,63 @@
+*** Settings ***
+Suite Setup    SuiteSetup
+Test Setup    TestSetup
+
+Default Tags    glue
+
+Resource    ../../../../../../resources/common/common_api.robot
+
+*** Test Cases ***
```
Comment (OKrivtsova) [inline tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot:9]: add a test for adding a gift card that was fully used (doe snot have money ay more (e.g. place order with it in precondition to max it out)

File: tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    ...  AND    Save value to a variable:    [data][id]    cart_id
+     When I send a POST request:    /carts/${cart_id}/cart-codes?include=vouchers,gift-cards    {"data": {"type": "cart-codes","attributes": {"code": "${random}"}}}
+    Then Response status code should be:    422
+    And Response should return error code:    3302
+    And Response should return error message:    "Cart code cant be added."
+
+
+
+
+adding_invalid_cart_code
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot]: done

File: tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+Resource    ../../../../../../resources/common/common_api.robot
+
+*** Test Cases ***
+ENABLER
+    TestSetup
+
+
+Create_git_card_code_with_invalid_access_token
+     [Setup]    Run Keyword   I set Headers:    Authorization=345A9
+    When I send a POST request:    /carts/cart_id/cart-codes?include=vouchers,gift-cards    {"data": {"type": "cart-codes","attributes": {"code": "${random}"}}}
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot]: done

File: tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    And Response reason should be:    Unauthorized
+    And Response should return error message:    Invalid access token.
+
+Create_git_card_code_without_access_token
+    When I send a POST request:    /carts/cart_id/cart-codes?include=vouchers,gift-cards    {"data": {"type": "cart-codes","attributes": {"code": "${random}"}}}
+    Then Response status code should be:    403
+    And Response should return error code:    002
+    And Response reason should be:    Forbidden 
+    And Response should return error message:    Missing access token.
+Create_git_card_code_with_invalid_cart_id
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot:29]: done

File: tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    ...  AND    Response status code should be:    201
+    ...  AND    Save value to a variable:    [data][id]    cart_id
+     When I send a POST request:    /carts/${cart_id}/cart-codes?include=vouchers,gift-cards    {"data": {"type": "cart-codes","attributes": {"code": "${random}"}}}
+    Then Response status code should be:    422
+    And Response should return error code:    3302
+    And Response should return error message:    "Cart code cant be added."
+
+
+
+
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot]: done

File: tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/positive.robot [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+Test Setup    TestSetup
+
+Default Tags    glue
+Resource    ../../../../../../resources/common/common_api.robot
+
+*** Test Cases ***
+ENABLER
+    TestSetup
+
+Add_gift_card_code_in_cart
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/positive.robot]: done

File: tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/positive.robot [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    And Response include element has self link:   gift-cards
+    And Response body parameter should not be EMPTY:   [included][0][attributes][code]
+    And Response body parameter should not be EMPTY:   [included][0][attributes][name]
+    And Response body parameter should be:   [included][0][attributes][value]    ${gift_card.amount}
+    And Response body parameter should be:   [included][0][attributes][currencyIsoCode]    EUR
+    And Response body parameter should be:   [included][0][attributes][actualValue]    ${gift_card.amount}
+    And Response body parameter should be:    [included][0][attributes][isActive]     True
+    And Response include should contain certain entity type:    gift-cards
+    And Response include element has self link:   gift-cards
+Delete_gift_card_code_from_the_guest_cart
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/positive.robot:114]: Done

File: tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/positive.robot [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+Add_gift_card_code_in_cart
+    [Setup]    Run Keywords    I get access token for the customer:    ${yves_second_user.email}
+    ...  AND    I set Headers:    Authorization=${token}
+    ...    AND     Create giftcode in Database:    ${random}    ${gift_card.amount}
+    ...  AND    I send a POST request:    /carts    {"data":{"type":"carts","attributes":{"priceMode":"${mode.gross}","currency":"${currency.eur.code}","store":"${store.de}","name": "${test_cart_name}-${random}"}}}
+    ...  AND    Response status code should be:    201
+    ...  AND    Save value to a variable:    [data][id]    cart_id
+    ...  AND    I send a POST request:    /carts/${cart_id}/items    {"data": {"type": "items","attributes": {"sku": "${concrete_available_product.with_label}","quantity": 1}}}
+    When I send a POST request:    /carts/${cart_id}/cart-codes?include=vouchers,gift-cards    {"data": {"type": "cart-codes","attributes": {"code": "${random}"}}}
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/positive.robot]: done

File: tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/positive.robot [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    And Response body parameter should be:   [included][0][attributes][actualValue]    ${gift_card.amount}
+    And Response body parameter should be:    [included][0][attributes][isActive]     True
+    And Response include should contain certain entity type:    gift-cards
+    And Response include element has self link:   gift-cards
+Delete_gift_card_code_from_the_guest_cart
+     [Setup]    I set Headers:    Content-Type=${default_header_content_type}    X-Anonymous-Customer-Unique-Id=${random}
+    Run Keywords    Create a guest cart:    ${random}   ${concrete_available_product.with_label}    1
+    ...    AND    Save value to a variable:    [data][id]    guestcart_id
+    ...    AND     Create giftcode in Database:    ${random}    ${gift_card.amount}
+     ...    AND    I send a POST request:    /guest-carts/${guestcart_id}/cart-codes?include=vouchers,gift-cards     {"data": {"type": "cart-codes","attributes": {"code": "${random}"}}}
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/positive.robot]: done

File: tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot (hunk 1 above)
Comment (khushal-rathore-metacube) [reply tests/api/suite_nonsplit/glue/cart_endpoints/cart_codes/negative.robot:9]: done

**Review by OKrivtsova — DISMISSED** [review]

**Review by OKrivtsova — APPROVED** [review]

**Pull request merged**